src/main/java/com/google/sps/data/FirestoreDB.java

```diff
@@ -0,0 +1,259 @@
+    ApiFuture addUserPropertyFuture =
+        userRef.update(Collections.singletonMap(nestedPropertyName, true));
+    DBUtils.blockOnFuture(addUserPropertyFuture);
+  }
+
+  public void deleteUserProperty(String userId, String objectId, String collection) {
+    DocumentReference userRef = DBUtils.user(userId);
+    String nestedPropertyName = DBUtils.getNestedPropertyName(collection, objectId);
+    ApiFuture removeUserPropertyFuture = userRef.update(nestedPropertyName, FieldValue.delete());
+    DBUtils.blockOnFuture(removeUserPropertyFuture);
+  }
+
+  public List<RecipeMetadata> getRecipesMatchingTags(
+      List<String> tagIds, SortingMethod sortingMethod) {
+    Query recipesQuery = recipesMatchingTags(tagIds, tagIds.iterator());
+    return getRecipeMetadataQuery(recipesQuery, sortingMethod);
+  }
+
+  public List<RecipeMetadata> getRecipesMatchingCreator(
+      String creatorId, SortingMethod sortingMethod) {
+    Query recipesQuery = DBUtils.recipes().whereEqualTo(Recipe.CREATOR_ID_KEY, creatorId);
+    return getRecipeMetadataQuery(recipesQuery, sortingMethod);
+  }
+  
+  public List<RecipeMetadata> getRecipesSavedBy(String userId, SortingMethod sortingMethod) {
+    List<String> saved_Ids = savedRecipeIds(userId);
+    return getRecipesMatchingIDs(saved_Ids, sortingMethod);
+  }
+
+  public List<RecipeMetadata> getRecipesMatchingIDs(List<String> Ids, SortingMethod sortingMethod) {
+    Query recipesQuery = DBUtils.recipes().whereIn(Recipe.ID_KEY, Ids);
+    return getRecipeMetadataQuery(recipesQuery, sortingMethod);
+  }
+  
+  public List<RecipeMetadata> getRecipeMetadataQuery(
+      Query recipesQuery, SortingMethod sortingMethod) {
+    switch (sortingMethod) {
+      case TOP:
+        recipesQuery = recipesQuery.orderBy(Recipe.VOTES_KEY, Query.Direction.DESCENDING);
+        break;
+      case NEW:
+        recipesQuery = recipesQuery.orderBy(Recipe.TIMESTAMP_KEY, Query.Direction.DESCENDING);
+        break;
+    }
+
+    ArrayList<RecipeMetadata> recipeList = new ArrayList<>();
+    QuerySnapshot querySnapshot = DBUtils.blockOnFuture(recipesQuery.get());
+
+    if (querySnapshot == null) {
+      return null;
+    }
+
+    for (DocumentSnapshot document : querySnapshot.getDocuments()) {
+      RecipeMetadata recipe = document.toObject(RecipeMetadata.class);
+      recipeList.add(recipe);
+    }
+    return recipeList;
+  }
+
+  public Query recipesMatchingTags(Iterable<String> tagIds, Iterator<String> iter) {
+    if (iter.hasNext()) {
+      String nextTag = iter.next();
+      return recipesMatchingTags(tagIds, iter).whereEqualTo("tags." + nextTag, true);
+    }
+    return DBUtils.recipes();
+  }
+
+  public List<String> savedRecipeIds(String userId) {
+    List<String> allRecipeIds = DBUtils.allRecipeIds();
+    ArrayList<String> savedRecipeIds = new ArrayList<String>();
+
+    for (String recipeID : allRecipeIds) {
+      if (isSavedRecipe(userId, recipeID)) {
+        savedRecipeIds.add(recipeID);
+      }
```

Comment:
Instead of filtering through all the recipes, don't we have a map in the user document that tells you which recipes are saved? We can just get the keyset of that and then we are done.